shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java

```diff
@@ -649,6 +651,69 @@ public void systemInsetDisplayCutoutSimple() {
     assertEquals(100, viewportMetricsCaptor.getValue().viewInsetTop);
   }
 
+  @SuppressWarnings("deprecation")
+  @Test
+  @Config(minSdk = 28)
+  public void onApplyWindowInsetsSetsDisplayCutouts() {
+    FlutterView flutterView = spy(new FlutterView(ctx));
+    assertEquals(0, flutterView.getSystemUiVisibility());
+    when(flutterView.getWindowSystemUiVisibility()).thenReturn(0);
+    when(flutterView.getContext()).thenReturn(ctx);
+
+    FlutterEngine flutterEngine = spy(new FlutterEngine(ctx, mockFlutterLoader, mockFlutterJni));
+    FlutterRenderer flutterRenderer = spy(new FlutterRenderer(mockFlutterJni));
+    when(flutterEngine.getRenderer()).thenReturn(flutterRenderer);
+
+    // When we attach a new FlutterView to the engine without any system insets,
+    // the viewport metrics default to 0.
+    flutterView.attachToFlutterEngine(flutterEngine);
+    ArgumentCaptor<FlutterRenderer.ViewportMetrics> viewportMetricsCaptor =
+        ArgumentCaptor.forClass(FlutterRenderer.ViewportMetrics.class);
+    verify(flutterRenderer).setViewportMetrics(viewportMetricsCaptor.capture());
+    assertEquals(0, viewportMetricsCaptor.getValue().viewPaddingTop);
+
+    // Then we simulate the system applying a window inset.
+    WindowInsets windowInsets = mock(WindowInsets.class);
+    DisplayCutout displayCutout = mock(DisplayCutout.class);
+    when(windowInsets.getDisplayCutout()).thenReturn(displayCutout);
+
+    List<Rect> boundingRects =
+        Arrays.asList(new Rect(0, 200, 300, 400), new Rect(150, 0, 300, 150));
+    when(displayCutout.getBoundingRects()).thenReturn(boundingRects);
+
+    // The following mocked methods are necessary to avoid a NullPointerException when calling
+    // onApplyWindowInsets, but are irrelevant to the behavior this test concerns.
+    Insets unusedInsets = Insets.of(100, 100, 100, 100);
+    // WindowInsets::getSystemGestureInsets was added in API 29, deprecated in API 30.
+    if (Build.VERSION.SDK_INT == 29) {
+      when(windowInsets.getSystemGestureInsets()).thenReturn(unusedInsets);
+    }
+    // WindowInsets::getInsets was added in API 30.
+    if (Build.VERSION.SDK_INT >= 30) {
+      when(windowInsets.getInsets(anyInt())).thenReturn(unusedInsets);
+    }
+    // DisplayCutout::getWaterfallInsets was added in API 30.
+    if (Build.VERSION.SDK_INT >= 30) {
+      when(displayCutout.getWaterfallInsets()).thenReturn(unusedInsets);
+    }
+    when(displayCutout.getSafeInsetTop()).thenReturn(100);
+    when(displayCutout.getSafeInsetLeft()).thenReturn(100);
+    when(displayCutout.getSafeInsetBottom()).thenReturn(100);
+    when(displayCutout.getSafeInsetRight()).thenReturn(100);
+
+    flutterView.onApplyWindowInsets(windowInsets);
+
+    verify(flutterRenderer, times(2)).setViewportMetrics(viewportMetricsCaptor.capture());
```

Comment:
Am I reading this correctly that we are updating the viewport metrics (`flutterRenderer.setViewportMetrics()`) twice when insets get applied? `sendViewportMetricsToFlutter()` only calls it once, right - do you know why we are seeing two invocations?

sorry if I'm misreading the test here, just want to be sure

Reply:
This is because it was already called once earlier in the test. I think there is a `clearInvocations` function I can call to avoid this; I'll have to try it

Reply:
got it, makes sense